### ISSUE TITLE
Interop failures between Heimdal master and AD regarding PAC (code 0x3c)

### ISSUE THREAD

**Issue by nicowilliams** [heimdal/heimdal#1091]
When a cross-realm TGT is issued by a Heimdal KDC running `master` branch code with a target realm that is an Active Directory realm, the AD KDC will respond with a `KRB-ERROR` with a generic error and will log a `0x3c` code which is [described](https://learn.microsoft.com/en-us/windows/security/threat-protection/auditing/event-4769) variously as:

```
Group membership has overloaded the PAC.
Multiple recent password changes have not propagated.
Crypto subsystem error caused by running out of memory.
SPN too long.
SPN has too many parts.
```

Workaround: set `disable_pac = true` in the Heimdal KDC's `krb5.conf` in `[realms]` under the KDC's realm.

**Comment by nicowilliams** [comment]
@iboukris @lhoward @josephsutton1 any clues?  Does this mean we should create a PAC large enough that the AD KDC can scribble SIDs onto it?

**Comment by nicowilliams** [comment]
Or maybe it's a problem with FAST?

**Comment by lhoward** [comment]
Not sure. @stevesyfuhs, the Heimdal KDC always issues PACs, even when not functioning as a Samba DC, in order to support constrained delegation and also to simplify the code paths for when it *is* functioning as a Samba DC.

This seems to be causing problems with cross-realm MIT trusts with Windows. Any ideas? Perhaps Windows isn't expecting this? The PACs Heimdal is issuing in this case omit `PAC_LOGON_INFO` but they do have `PAC_SERVER_CHECKSUM`, `PAC_PRIVSVR_CHECKSUM`, `PAC_LOGON_NAME`,`PAC_UPN_DNS_INFO`, `PAC_ATTRIBUTES_INFO`, and `PAC_TICKET_CHECKSUM` as far as I can tell from reading the code.

**Comment by nicowilliams** [comment]
I've tried excluding `PAC_UPN_DNS_INFO`; that did not help.

**Comment by josephsutton1** [comment]
Could it be because `PAC_LOGON_INFO` is missing? Samba at least requires that buffer to be present in the PACs it receives.

**Comment by lhoward** [comment]
Quite possible. Windows seems to be OK without `PAC_LOGON_INFO` in the same domain (i.e. when pretending to be a RODC à la Azure AD or TktBridgeAP, including only `PAC_REQUESTOR_SID`). But likely the MIT cross-realm code paths are different in Windows.

**Comment by SteveSyfuhs** [comment]
It does seem odd to issue a PAC without the logon info but with everything else.

Unfortunately, when the AD KDC returns a generic error that tends to mean one of two things: out-of-memory (exit asap to return a minimally viable error), or we landed at a point that is totally confusing to us and we never expected it to get there.
________________________________

**Comment by nicowilliams** [comment]
Heimdal's TGS client really wants FAST now, but I disabled it with a hammer and still x-realm to AD fails, so I assume it's not FAST.  (I should push a commit to allow disabling of FAST for TGS, and I should push a commit to complete the name canon rule functionality that wants FAST.)

**Comment by nicowilliams** [comment]
> Perhaps we should always omit the PAC for cross-realm TGTs unless we know we're using Samba.

I thought AD was soon to _require_ PACs even for the MIT realm x-realm case.  Am I hallucinating that?  We do have a way to mark principals in the HDB as no-authdata-reqd, which is a way of saying "no PAC".

**Comment by nicowilliams** [comment]
@SteveSyfuhs ah, so:

- what PAC buffer types does AD want?
- will AD always accept x-realm `krbtgt`s from non-AD (MIT) realms or will it soon insist on them?  maybe only for x-realm SFU?
- does AD always accept a `PAC_LOGON_NAME` w/ a realm name included, or only for x-realm SFU?
- does receiving a cross-realm TGS-REQ with from a MIT/Heimdal realm w/ a minimal PAC in the x-realm `krbtgt` allow `altSecIds` to continue working or does it break `altSecIds`?

**Comment by SteveSyfuhs** [comment]
I didn't think we were planning on putting that requirement on MIT trusts. Obviously, we'd prefer it if we could, but that's a major breaking change.
________________________________

**Comment by nicowilliams** [comment]
> I didn't think we were planning on putting that requirement on MIT trusts. Obviously, we'd prefer it if we could, but that's a major breaking change.

Ahh, that's a relief!  So root causing and fixing this is not as urgent as I thought.  *phew*

**Comment by SteveSyfuhs** [comment]
Let me confirm what our specific requirements around that will be. This could still be an annoying thing to thread one way or another and I don't want to give you the wrong information.
________________________________

**Comment by iboukris** [comment]
Ouch, as Luke pointed out the mini PAC was introduced for S4U2proxy support, to avoid duplicate code and logic. If it causes interop issues perhaps another auth-data type number could be used.

CC: @greghudson

**Comment by nicowilliams** [comment]
@iboukris I don't think we'll have to do that.  This will be my patch:

```
commit 70df93b08c08a07392b968b90cca64f86e455433 (HEAD -> master)
Author: Nicolas Williams <nico@twosigma.com>
Date:   Wed May 17 21:17:13 2023 -0500

    kdc: Honor no-auth-data-reqd on cross-real TGTs

    Nowadays we use PACs instead of AD-SIGNEDPATH, so we want a PAC on every
    TGT, but we don't necessarily want PACs on cross-realm TGTs.

    Specifically, we don't interop well yet with AD when issuing cross-realm
    TGTs with AD realms as the destination realm (see #1091).

diff --git a/kdc/misc.c b/kdc/misc.c
index 5227eb42f39..8c6a7641a4c 100644
--- a/kdc/misc.c
+++ b/kdc/misc.c
@@ -248,14 +248,20 @@ _kdc_verify_checksum(krb5_context context,
  * tickets, policy is governed by whether the client explicitly requested
  * a PAC be omitted when requesting a TGT, or if the no-auth-data-reqd
  * flag is set on the service principal entry.
+ *
+ * However, when issuing a cross-realm TGT to an AD realm our PAC might not
+ * interoperate correctly.  Therefore we honor the no-auth-data-reqd HDB entry
+ * flag on cross-realm TGTs.
  */

 krb5_boolean
 _kdc_include_pac_p(astgs_request_t r)
 {
-    if (krb5_principal_is_krbtgt(r->context, r->server->principal))
+    /* Always include a PAC in root TGTs */
+    if (krb5_principal_is_krbtgt(r->context, r->server->principal) &&
+        krb5_principal_is_root_krbtgt(r->context, r->server->principal))
        return TRUE;
-    else if (r->server->flags.no_auth_data_reqd)
+    if (r->server->flags.no_auth_data_reqd)
        return FALSE;

     return !!(r->pac_attributes & (KRB5_PAC_WAS_REQUESTED | KRB5_PAC_WAS_GIVEN_IMPLICITLY));
```

which will allow us to mark cross-realm `krbtgt` principals so that PACs are not included in cross-realm TGTs to those target realms.

We can then begin _leisurely_ working with Microsoft to make sure that the cross-realm TGTs we issue have PACs that interop with Active Directory.  That work may include Heimdal realms becoming SID authorities that can at least assign a distinct RID to every new local principal, who knows.  I'm sure it will be fun enough.  But we won't have to do it right away :)

**Comment by iboukris** [comment]
@nicowilliams, right, for Heimdal that should be fine, in MIT though we need the PAC in xrealm tickets as well for RBCD.

**Comment by nicowilliams** [comment]
@iboukris so is this working for MIT?  If so, what's MIT doing differently from Heimdal that makes it work?

**Comment by iboukris** [comment]
@nicowilliams, idk, I'd guess MIT would have the same problem (my knowledge is quite rusty tbh).

**Comment by nicowilliams** [comment]
@frozencemetery have you encountered anything like this with MIT Kerberos?

**Comment by nicowilliams** [comment]
I'm told I should ping @abbra rather than @frozencemetery.

**Comment by abbra** [comment]
Last year I saw issues when AD DC did respond to our (freeipa) s4u requests when we used sha2-based checksum. This happened when we started to default to RFC8009 encryption types for krbtgts. AD DCs do not support that, dochelp@ confirmed it was rejection due to unknown checksum type. This was silently fixed in November 2022 security update. 

So I would ask what Windows version is used for tests and whether Heimdal also defaults to sha2-based encryption types in your tests.

**Comment by iboukris** [comment]
Note that freeipa does generate a full PAC like samba though.
